### PR TITLE
Add x-oro token contract & hardhat verify contracts information

### PR DESCRIPTION
**Pull request by dqtkien**

### CODE REVIEW

File: contracts/orochi/XOroV2.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.19;
+
+import '@openzeppelin/contracts/token/ERC1155/ERC1155.sol';
+import '@openzeppelin/contracts/access/Ownable.sol';
+
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol:6]:
```suggestion
error AccessDenied();
```

File: contracts/orochi/XOroV2.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+contract XOroV2 is ERC1155, Ownable {
+  uint8 public constant TOKEN_ID = 1;
+  constructor() ERC1155('https://metadata.orochi.network/x-oro-v2/{id}.json') {}
+
+  function mint(address to, uint256 amount) public onlyOwner {
+    _mint(to, TOKEN_ID, amount, '');
+  }
+
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol:22]:
```suggestion
modifier accessDenied() {
   revert AccessDenied();
   _;
}
```

File: contracts/orochi/XOroV2.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  function balance(address account) public view virtual returns (uint256) {
+    return super.balanceOf(account, TOKEN_ID);
+  }
+
+  function safeTransferFrom(address from, address to, uint256 id, uint256 amount, bytes memory data) public override {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
  function safeTransferFrom(address from, address to, uint256 id, uint256 amount, bytes memory data) public override accessDenied {
```

File: contracts/orochi/XOroV2.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    return super.safeTransferFrom(from, to, id, amount, data);
+  }
+
+  function safeBatchTransferFrom(
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  ) public override {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
  ) public override accessDenied {
```

File: contracts/orochi/XOroV2.sol [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    require(false, 'Transfer not allowed');
+    return super.safeBatchTransferFrom(from, to, ids, amounts, data);
+  }
+
+  function setApprovalForAll(address operator, bool approved) public virtual override {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
  function setApprovalForAll(address operator, bool approved) public virtual override accessDenied {
```

File: contracts/orochi/XOroV2.sol (hunk 1 above)
Comment (dqtkien) [reply contracts/orochi/XOroV2.sol:6]: Fixed

File: contracts/orochi/XOroV2.sol (hunk 2 above)
Comment (dqtkien) [reply contracts/orochi/XOroV2.sol:22]: Fixed

File: contracts/orochi/XOroV2.sol (hunk 5 above)
Comment (dqtkien) [reply contracts/orochi/XOroV2.sol]: Fixed

File: contracts/orochi/XOroV2.sol [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  uint8 public constant TOKEN_ID = 1;
+
+  modifier accessDenied() {
+    revert AccessDenied();
+    _;
+  }
+
+  constructor() ERC1155('https://metadata.orochi.network/x-oro-v2/{id}.json') {}
+
+  function mint(address to, uint256 amount) public onlyOwner {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]: Public isn't secure enough

Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
  function mint(address to, uint256 amount) external onlyOwner {
```

File: contracts/orochi/XOroV2.sol [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    _mint(to, TOKEN_ID, amount, '');
+  }
+
+  function batchMint(uint256[] calldata packedData) public onlyOwner {
+    for (uint i = 0; i < packedData.length; i += 1) {
+      (uint96 amount, address beneficiary) = _unpack(packedData[i]);
+      mint(beneficiary, amount);
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
       _mint(to, TOKEN_ID, amount, '');
```

File: contracts/orochi/XOroV2.sol [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    _;
+  }
+
+  constructor() ERC1155('https://metadata.orochi.network/x-oro-v2/{id}.json') {}
+
+  function mint(address to, uint256 amount) public onlyOwner {
+    _mint(to, TOKEN_ID, amount, '');
+  }
+
+  function batchMint(uint256[] calldata packedData) public onlyOwner {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]:
```suggestion
  function batchMint(uint256[] calldata packedData) external onlyOwner {
```

File: contracts/orochi/XOroV2.sol [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    for (uint i = 0; i < packedData.length; i += 1) {
+      (uint96 amount, address beneficiary) = _unpack(packedData[i]);
+      mint(beneficiary, amount);
+    }
+  }
+
+  function balance(address account) public view virtual returns (uint256) {
```
Comment (chiro-hiro) [inline contracts/orochi/XOroV2.sol]: Is `balance()` a correct method?

File: contracts/orochi/XOroV2.sol (hunk 6 above)
Comment (dqtkien) [reply contracts/orochi/XOroV2.sol]: Fixed

Comment (dqtkien) [reply contracts/orochi/XOroV2.sol]: Fixed

File: contracts/orochi/XOroV2.sol (hunk 9 above)
Comment (dqtkien) [reply contracts/orochi/XOroV2.sol]: At the begining, I add this function to delegate balanceOf without adding TOKEN_ID parameter. But now I think it is unnecessary. Deleted.

**Review by chiro-hiro — APPROVED** [review]

**Pull request merged**